espnet/nets/pytorch_backend/conformer/attention.py

```diff
@@ -0,0 +1,190 @@
+        :param torch.Tensor value: (batch, time2, size)
+        :param torch.Tensor mask: (batch, time1, time2)
+        :param torch.nn.Dropout dropout:
+        :return torch.Tensor: attentined and transformed `value` (batch, time1, d_model)
+             weighted by the query dot key attention (batch, head, time1, time2)
+        """
+        n_batch = query.size(0)
+        q = self.linear_q(query).view(n_batch, -1, self.h, self.d_k)
+        k = self.linear_k(key).view(n_batch, -1, self.h, self.d_k)
+        v = self.linear_v(value).view(n_batch, -1, self.h, self.d_k)
+        q = q.transpose(1, 2)  # (batch, head, time1, d_k)
+        k = k.transpose(1, 2)  # (batch, head, time2, d_k)
+        v = v.transpose(1, 2)  # (batch, head, time2, d_k)
+
+        scores = torch.matmul(q, k.transpose(-2, -1)) / math.sqrt(
+            self.d_k
+        )  # (batch, head, time1, time2)
+        if mask is not None:
+            mask = mask.unsqueeze(1).eq(0)  # (batch, 1, time1, time2)
+            min_value = float(
+                numpy.finfo(torch.tensor(0, dtype=scores.dtype).numpy().dtype).min
+            )
+            scores = scores.masked_fill(mask, min_value)
+            self.attn = torch.softmax(scores, dim=-1).masked_fill(
+                mask, 0.0
+            )  # (batch, head, time1, time2)
+        else:
+            self.attn = torch.softmax(scores, dim=-1)  # (batch, head, time1, time2)
+
+        p_attn = self.dropout(self.attn)
+        x = torch.matmul(p_attn, v)  # (batch, head, time1, d_k)
+        x = (
+            x.transpose(1, 2).contiguous().view(n_batch, -1, self.h * self.d_k)
+        )  # (batch, time1, d_model)
+        return self.linear_out(x)  # (batch, time1, d_model)
+
+
+class RelPositionMultiHeadedAttention(nn.Module):
+    """Multi-Head Attention layer with relative position encoding.
+
+    :param int n_head: the number of head s
+    :param int n_feat: the number of features
+    :param float dropout_rate: dropout rate
+
+    """
+
+    def __init__(self, n_head, n_feat, dropout_rate):
+        """Construct an RelPositionMultiHeadedAttention object."""
+        super(RelPositionMultiHeadedAttention, self).__init__()
+        assert n_feat % n_head == 0
+        # We assume d_v always equals d_k
+        self.d_k = n_feat // n_head
+        self.h = n_head
+        self.linear_q = nn.Linear(n_feat, n_feat)
+        self.linear_k = nn.Linear(n_feat, n_feat)
+        self.linear_v = nn.Linear(n_feat, n_feat)
+        # linear transformation for positional ecoding
+        self.linear_pos = nn.Linear(n_feat, n_feat, bias=False)
+        # these two hyperparameters are used in matrix c and matrix d
+        # as described in https://arxiv.org/abs/1901.02860 Section 3.3
+        self.pos_bias_u = nn.Parameter(torch.Tensor(self.h, self.d_k))
+        self.pos_bias_v = nn.Parameter(torch.Tensor(self.h, self.d_k))
+        torch.nn.init.kaiming_uniform_(self.pos_bias_u, a=math.sqrt(5))
```

Comment:
What is the reference for this part?

Reply:
The reference is the Transformer-XL paper. They introduce two learnable bias named u and v for computing rel. pos ecoding. I named these two tensors as described in the paper. Since `bias_u` and `bias_v` are not easy to understand, I just cite the paper here. Should we change another name? Thanks so much for your correction and advice.

Reply:
My question is about initialization. Original Transformer-XL initialized parameters by sampling values from normal distribution with std 0.02. Where did you find the reference to initialize these bias parameters by sampling from kaiming_uniform?

Reply:
I saw the initialization of the learnable bias is `self.r_w_bias = nn.Parameter(torch.Tensor(self.n_head, self.d_head))` in the official implementation of transformer-xl (https://github.com/hirofumi0810/neural_sp/blob/master/neural_sp/models/seq2seq/encoders/transformer.py#L165). I found the bias values of this method can be very large, such as `6.0630e+33, 4.5836e-41`. So I used the same initialization function as `torch.nn.Linear.reset_parameters()`. I am not sure about the influence between different initialization functions, maybe I should change it to `torch.nn.init.xavier_normal_`.